ReactCommon/butter/function.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <butter/butter.h>
+
+#if (defined(_WIN32) && __has_include(<functional/functor.h>))
+
+#include <functional/functor.h>
+
+#elif defined(BUTTER_USE_FOLLY_CONTAINERS)
+
+#include <folly/Function.h>
+
+#else
+
+#include <functional>
+
+#endif
+
+namespace facebook {
+namespace butter {
+
+#if (defined(_WIN32) && __has_include(<functional/functor.h>))
+
+template <typename... Ts>
+using function = Mso::Functor<Ts...>;
+
+#elif defined(BUTTER_USE_FOLLY_CONTAINERS)
+
+template <typename... Ts>
+using function = folly::Function<Ts...>;
+
+#else
+
+template <typename... Ts>
+using function = std::function<Ts...>;
+
+#define BUTTER_USE_STD_FUNCTION = 1
```

Comment:
It may be worth to add `BUTTER_FUNCTION_CAN_MOVE` macro (or some other name along these lines) to indicate if the butter function supports move semantic. It should help to `#ifdef` user code in case if we need to capture move-only parameters. 
Or maybe even define a special function/type that could help avoiding checking `BUTTER_FUNCTION_CAN_MOVE` for  each butter::function use.

Reply:
I think I leave that out for a future change ;-)
Feel fee to PR that.

I need 'BUTTER_USE_STD_FUNCTION' right now for react/bridging/Function.h to gate this template

```
#ifndef BUTTER_USE_STD_FUNCTION
template <typename R, typename... Args>
struct Bridging<std::function<R(Args...)>>
    : Bridging<butter::function<R(Args...)>> {};
#endif
```

The problem is if butter::function is already a std::function > this template throws a compile error.

- Do you know a smarter way to conditionally disable this template:
```
template <typename R, typename... Args>
struct Bridging<std::function<R(Args...)>>
    : Bridging<butter::function<R(Args...)>> {};
```
w/o needing to introduce a `BUTTER_USE_STD_FUNCTION` flag?

Reply:
It is possible to do if we change the `Bridging` to have extra `Cond` template argument `template <typename T, typename Cond = void> struct Bridging;`. No other code needs to change, but for the `std::function` specialization we can add a condition to enable it only if `std::function != butter::function`:
```
template <typename R, typename... Args>
struct Bridging<std::function<R(Args...)>,
  std::enable_if_t<!std::is_same_v<std::function<R(Args...)>, butter::function<R(Args...)>>>>
  : Bridging<butter::function<R(Args...)>> { ... };
```